### PR TITLE
8331418: ZGC: generalize barrier liveness logic

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset generalizes the logic to analyze, declare, and communicate which registers are live at a C2 barrier stub so that it can be used by other collectors than ZGC adopting the late barrier expansion model (including G1 in the near future, see [JEP 475](https://openjdk.org/jeps/475)).

The main changes are:

- Make it possible to compute register liveness information before (live-in) or after (live-out) each barrier, and let the collector choose by implementing `BarrierSetC2State::needs_livein_data()`.

- Generalize the interface with which collectors declare which registers must be additionally preserved across barrier runtime calls, adding the methods `BarrierStubC2::preserve(Register r)` and `BarrierStubC2::dont_preserve(Register r)`.

- Simplify the interface with which platform-specific logic computes which registers to preserve across barrier runtime calls, replacing the calls to `BarrierStubC2::result()` and `BarrierStubC2::live()` with a single call to `BarrierStubC2::preserve_set()`.

#### Testing

- tier1-5 (windows-x64, linux-x64, linux-aarch64, macosx-x64, macosx-aarch64; release and debug mode).
- tier1-7 (linux-x64, linux-aarch64; release and debug mode; ZGC tests only) with [an additional patch](https://github.com/openjdk/jdk/commit/4d4e743d8f4cddd5288cee1d69c70fe2b9bea066) that exercises the spilling and restoring logic by forcing ZGC read barriers to always take the slow path and clearing all general-purpose save-on-call registers upon the slow path's runtime call.
- Build with `make hotspot` (linux-riscv64-debug, linux-ppc64le-debug). @RealFYang, @TheRealMDoerr: could you please test and review the riscv and ppc changes? Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331418](https://bugs.openjdk.org/browse/JDK-8331418): ZGC: generalize barrier liveness logic (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [c0fc66de](https://git.openjdk.org/jdk/pull/19026/files/c0fc66deb654a9b930a7b7cf1a7e7fa093739027)
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**) ⚠️ Review applies to [c0fc66de](https://git.openjdk.org/jdk/pull/19026/files/c0fc66deb654a9b930a7b7cf1a7e7fa093739027)
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19026/head:pull/19026` \
`$ git checkout pull/19026`

Update a local copy of the PR: \
`$ git checkout pull/19026` \
`$ git pull https://git.openjdk.org/jdk.git pull/19026/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19026`

View PR using the GUI difftool: \
`$ git pr show -t 19026`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19026.diff">https://git.openjdk.org/jdk/pull/19026.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19026#issuecomment-2086593657)

### CODE REVIEW

File: src/hotspot/share/gc/shared/c2/barrierSetC2.cpp [hunk 1]
```diff
@@ -108,6 +109,33 @@ Label* BarrierStubC2::continuation() {
...
+void BarrierStubC2::dont_preserve(Register r) {
+  const VMReg vm_reg = r->as_VMReg();
+  assert(vm_reg->is_Register(), "r must be a general-purpose register");
+  OptoReg::Name reg = OptoReg::as_OptoReg(vm_reg);
+  // Subtract not only reg, but also all related OptoRegs that are sub-registers
+  // of the same general-purpose, processor register (e.g. {R11, R11_H} for r11
+  // in aarch64). We assume that all related OptoRegs have contiguous indices.
+  while (OptoReg::is_reg(reg)) {
+    const VMReg vm_reg = OptoReg::as_VMReg(reg);
+    if (!(vm_reg->is_Register()) || vm_reg->as_Register() != r) {
```
Comment (TheRealMDoerr) [inline src/hotspot/share/gc/shared/c2/barrierSetC2.cpp:127]: This doesn't work on PPC64: We run into "assert(is_Register() && is_even(value())) failed: even-aligned GPR name" (vmreg_ppc.hpp:54). Calling `as_Register()` is only supported for the even ones.

Comment (robcasloz) [reply src/hotspot/share/gc/shared/c2/barrierSetC2.cpp:127]: Thank you Martin for trying out the patch and for the suggestion, will test a solution based on `is_concrete()` and push the changes later, if it works.

Comment (robcasloz) [reply src/hotspot/share/gc/shared/c2/barrierSetC2.cpp]: Done (commit https://github.com/openjdk/jdk/pull/19026/commits/c0fc66deb654a9b930a7b7cf1a7e7fa093739027). @TheRealMDoerr please let me know if this works on PPC64.

Comment (TheRealMDoerr) [reply src/hotspot/share/gc/shared/c2/barrierSetC2.cpp]: Yes, this works on PPC64. Thanks!

**Comment by TheRealMDoerr** [comment]
Can we change `_barrier_set_state` (https://github.com/openjdk/jdk/blob/a024eed7384828643e302f021a253717f53e3778/src/hotspot/share/opto/compile.hpp#L364) from `void*` to `BarrierSetC2State*` and remove the casts?

**Comment by robcasloz** [comment]
> Can we change `_barrier_set_state` (
> 
> https://github.com/openjdk/jdk/blob/a024eed7384828643e302f021a253717f53e3778/src/hotspot/share/opto/compile.hpp#L364
> 
> ) from `void*` to `BarrierSetC2State*` and remove the casts?

Thanks for the suggestion, this would be a nice improvement, however it would be fairly pervasive (I sketched it it in https://github.com/openjdk/jdk/commit/cf5c1587e0ea90a8b3de4c70e0a2bf6ba4158f15), so I think it would be better to apply it as a separate RFE, perhaps after [non-generational ZGC is removed](https://openjdk.org/jeps/474) for simplicity. What do you think?

**Review by TheRealMDoerr — APPROVED** [review]
I haven't thought about future usages of `BarrierSetC2State::needs_livein_data()`. I guess it's intended for G1. Otherwise, LGTM.

File: src/hotspot/share/gc/shared/c2/barrierSetC2.cpp [hunk 2]
```diff
@@ -87,15 +87,16 @@ static BarrierSetC2State* barrier_set_state() {
...
-
 RegMask& BarrierStubC2::live() const {
   return *barrier_set_state()->live(_node);
 }
 
+BarrierStubC2::BarrierStubC2(const MachNode* node)
+  : _node(node),
+    _entry(),
+    _continuation(),
+    _preserve(live()){}
```
Comment (xmas92) [inline src/hotspot/share/gc/shared/c2/barrierSetC2.cpp:98]:
```suggestion
    _preserve(live()) {}
```

**Comment by robcasloz** [comment]
Thanks for reviewing, Martin! I reported your suggested refactoring here: https://bugs.openjdk.org/browse/JDK-8331623.

> I haven't thought about future usages of `BarrierSetC2State::needs_livein_data()`. I guess it's intended for G1.

That's correct, it is primarily intended for G1. But ZGC could also benefit, in the future, from using live-out instead of live-in data in the spilling logic. The current solution is slightly over-conservative in that it might spill some registers unnecessarily.

File: src/hotspot/share/gc/shared/c2/barrierSetC2.cpp [hunk 3]
```diff
@@ -850,6 +873,14 @@ void BarrierSetC2::compute_liveness_at_stubs() const {
     for (int i = block->number_of_nodes() - 1; i >= 0; --i) {
       const Node* const node = block->get_node(i);
 
+      // If this node tracks out-liveness, update it
+      if (!bs_state->needs_livein_data()) {
+        RegMask* const regs = bs_state->live(node);
+        if (regs != NULL) {
```
Comment (xmas92) [inline src/hotspot/share/gc/shared/c2/barrierSetC2.cpp:879]:
```suggestion
        if (regs != nullptr) {
```

File: src/hotspot/share/gc/shared/c2/barrierSetC2.cpp [hunk 4]
```diff
@@ -873,10 +904,12 @@ void BarrierSetC2::compute_liveness_at_stubs() const {
...
       }
 
-       // If this node tracks liveness, update it
-      RegMask* const regs = barrier_set_state()->live(node);
-      if (regs != NULL) {
-        regs->OR(new_live);
+      // If this node tracks in-liveness, update it
+      if (bs_state->needs_livein_data()) {
+        RegMask* const regs = bs_state->live(node);
+        if (regs != NULL) {
```
Comment (xmas92) [inline src/hotspot/share/gc/shared/c2/barrierSetC2.cpp:910]:
```suggestion
        if (regs != nullptr) {
```

**Review by xmas92 — APPROVED** [review]
lgtm.

A few nits.

**Comment by robcasloz** [comment]
> lgtm.
> 
> A few nits.

Thanks for reviewing and for the style suggestions, Axel!

**Review by RealFYang — APPROVED** [review]
@robcasloz : This also tests good on linux-riscv64 platform. LGTM. Thanks for the ping!

**Comment by robcasloz** [comment]
> @robcasloz : This also tests good on linux-riscv64 platform. LGTM. Thanks for the ping!

Thanks for testing!

**Review by fisk — APPROVED** [review]
Looks good!

**Comment by robcasloz** [comment]
> Looks good!

Thanks for reviewing, Erik!

**Pull request closed without merging**